R/PseudoBulk.R

```diff
@@ -352,3 +356,140 @@ CreateStudyWideBarPlot <- function(pairwise_de_results, pvalue_threshold = 0.05,
+      gate <- logical_dataframe[logic_row_index,"field_logic"]
+      criterion <- logical_dataframe[logic_row_index,"criteria"]
+      #set up logic gate
+      if(get(gate)(
+        positive_contrast[field_to_check] == criterion,
+        negative_contrast[field_to_check] == criterion)){
+        
+        gates_satisfied <- gates_satisfied + 1
+        #it is possible to fully define your study design within logical_dataframe without using the require_identical_fields logic as a shortcut.
+        if (use_require_identical_logic == T){
+          #check if the contrast passed all of the logic gates, if so, test for require_identical fields.
+          if (gates_satisfied == nrow(logical_dataframe)){
+            
+            #reset counter for matching require_identical fields.
+            require_identical_fields_satisfied <- 0
+            for (require_identical in require_identical_fields){
+              #keep track of how many require_identical fields are equal to each other.
+              if(positive_contrast[require_identical] == negative_contrast[require_identical]){
+                require_identical_fields_satisfied <- require_identical_fields_satisfied + 1
+              }
+              #if all of the require_identical fields are require_identical keep the contrast.
+              if (require_identical_fields_satisfied == length(require_identical_fields)){
+                filtered_contrast_indices <- c(filtered_contrast_indices, row_index)
+              }
+            }
+          }
+        } else{
+          #if all of the gates are satisfied and the require_identical logic isn't used, keep the contrast.
+          if (gates_satisfied == nrow(logical_dataframe)){
+            filtered_contrast_indices <- c(filtered_contrast_indices, row_index)
+          }
+        }
+      }
+    }
+  }
+  #filter the contrasts using the indices that satisfy the logic gates/require_identical fields.
+  contrasts <- contrasts[filtered_contrast_indices,]
+  
+  #initialize a dataframe to store the filtered contrasts.
+  filtered_contrasts_dataframe <- data.frame()
+  #iterate through the filtered contrasts and coerce the matrix into a data frame. 
+  for(contrast_index in 1:nrow(contrasts)){
+    #split the contrasts by _ and coerce into a dataframe. 
+    temporary_data_frame <- t(as.data.frame(unlist(strsplit(contrasts[contrast_index,], split = "_"))))
+    #assign the column names of the dataframe via the ordering supplied by contrast_columns.
+    colnames(temporary_data_frame) <- c(paste0("positive_contrast_", contrast_columns), 
+                                        paste0("negative_contrast_", contrast_columns))
+    rownames(temporary_data_frame) <- contrast_index
+    filtered_contrasts_dataframe <- rbind(filtered_contrasts_dataframe, temporary_data_frame)
+  }
+  
+  write.table(filtered_contrasts_dataframe, file = filtered_contrasts_output_file, row.names = F, col.names = T)
```

Comment:
The output is now a dataframe-like table with column names like "positive_contrast_Timepoint" and "negative_contrast_Timepoint" which I'll parse in the next function that actually runs the contrasts.